### PR TITLE
Cleanup

### LINKED ISSUE #2: ConnectionObserverAdapter not needed, `stateAsFlow()` can be used
[ConnectionObserverAdapter](https://github.com/NordicSemiconductor/Android-nRF-Wi-Fi-Provisioner/blob/1d10bd6c577a21718511c932f4b8ac63ee83e1e3/lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/internal/ConnectionObserverAdapter.kt) class is redundant.

The `ble-ktx` library already provides state as Flow using `stateAsFlow()` extension:
https://github.com/NordicSemiconductor/Android-BLE-Library/blob/60b8b90ef5e276b0c6de005ada993ec6b9a52d55/ble-ktx/src/main/java/no/nordicsemi/android/ble/ktx/BleManagerExt.kt#L43-L56

### PR DESCRIPTION
**Pull request by philips77**
This is a combo PR fixing the following issues:
1. Packet refactoring
2. Making internal types internal in the library while at the same time moving public types out from *internal* package
3. Improvements in BLE library
   * Response types now decode themselves
4. Fixing #2 by removing the type and providing mapping

### CODE REVIEW

File: lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/ProvisionerRepository.kt [hunk 1]
```diff
@@ -72,56 +73,58 @@ interface ProvisionerRepository {
...
-     * @throws [ResponseErrorException] - when the IoT reports result different that success
+     * Stops scanning for available Wi-Fi networks. Should be called after [startScan].
+     *
+     * @throws [ResponseErrorException] - when the device reports result different that success.
      */
     suspend fun stopScan()
 
     /**
-     * Provision the connected DK with data obtained from [startScan] + password.
+     * Provisions the connected device with data given Wi-Fi configuration.
```
Comment (sylwester-zielinski) [inline lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/ProvisionerRepository.kt]: There is sth grammatically incorrect with this sentence.

File: lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/ProvisionerRepositoryImpl.kt [hunk 2]
```diff
@@ -34,60 +34,45 @@ package no.nordicsemi.wifi.provisioner.library
...
-    }
+    override suspend fun readVersion(): VersionDomain = manager?.getVersion()?.toDomain()
+        ?: throw IllegalStateException(MANAGER_NOT_INITIALIZED)
 
-    override suspend fun getStatus(): DeviceStatusDomain {
-        val status = manager?.getStatus()
-        return status?.toDomain()!!
-    }
+    override suspend fun getStatus(): DeviceStatusDomain = manager?.getStatus()?.toDomain()
+        ?: throw IllegalStateException(MANAGER_NOT_INITIALIZED)
```
Comment (sylwester-zielinski) [inline lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/ProvisionerRepositoryImpl.kt]: Maybe we can du some property: val forceManager = manager ?: throw IllegalStateException(MANAGER_NOT_INITIALIZED). 
We would remove many duplication of throw and ? operators.

File: lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/ProvisionerRepositoryImpl.kt [hunk 3]
```diff
@@ -98,14 +83,14 @@ class ProvisionerRepositoryImpl internal constructor(
...
 
-    private fun <T> runTask(block: suspend () -> T): Flow<Resource<T>> {
-        return flow { emit(Resource.createSuccess(block())) }
-            .onStart { emit(Resource.createLoading()) }
-            .catch {
-                it.printStackTrace()
-                emit(Resource.createError(it))
-            }
-    }
+//    private fun <T> runTask(block: suspend () -> T): Flow<Resource<T>> {
```
Comment (sylwester-zielinski) [inline lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/ProvisionerRepositoryImpl.kt]: Why comment? Remove it.

File: lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/internal/ProvisionerBleManager.kt [hunk 4]
```diff
@@ -105,8 +92,26 @@ internal class ProvisionerBleManager(
                 .useAutoConnect(false)
                 .retry(3, 100)
                 .suspend()
-            createBondInsecure().suspend()
-            dataHolder.status
+            createBondInsecure()
+                .suspend()
+            stateAsFlow().map {
```
Comment (sylwester-zielinski) [inline lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/internal/ProvisionerBleManager.kt:97]: I would remove ConnectionStatus and use ConnectionState instead.

File: lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/internal/ProvisionerBleManager.kt [hunk 5]
```diff
@@ -283,25 +272,23 @@ internal class ProvisionerBleManager(
...
     }
 
-    private fun verifyResponseSuccess(response: ByteArray) {
-        val status = Response.ADAPTER.decode(response).status
-        if (status != Status.SUCCESS) {
-            throw createResponseError(status)
-        }
+    private fun verifyResponseSuccess(response: ResponsePacket) {
+        response.value.status
+            .takeUnless { it == Status.SUCCESS }
```
Comment (sylwester-zielinski) [inline lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/internal/ProvisionerBleManager.kt:320]: Master of Kotlin!

File: lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/internal/response/ResponsePacket.kt [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import no.nordicsemi.android.wifi.provisioning.Response
+
+internal class ResponsePacket : ProfileReadResponse() {
+    lateinit var value: Response
+
+    override fun onDataReceived(device: BluetoothDevice, data: Data) {
+        super.onDataReceived(device, data)
+
+        try {
+            value = Response.ADAPTER.decode(data.value ?: byteArrayOf())
```
Comment (sylwester-zielinski) [inline lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/internal/response/ResponsePacket.kt:46]: I think this try-catch will not propagate exception to UI. This is not good. Please fix.

**Review by sylwester-zielinski — CHANGES_REQUESTED** [review]

File: lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/ProvisionerRepository.kt (hunk 1 above)
Comment (philips77) [reply lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/ProvisionerRepository.kt]: data -> the

File: lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/internal/response/ResponsePacket.kt (hunk 6 above)
Comment (philips77) [reply lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/internal/response/ResponsePacket.kt:46]: Indeed, there's an issue, but not that there would be no exception.
`suspendForValidResponse()` would throw `InvalidDataException` instead, but the problem is, that this exception is defined in the BLE library. A module using the provisioning library does not need to depend on BLE library, so there can be issues with catching this exception, as its type is not exported by the lib.

File: lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/internal/ProvisionerBleManager.kt (hunk 4 above)
Comment (philips77) [reply lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/internal/ProvisionerBleManager.kt:97]: Here's the similar issue to what I explained below. The `ConnectionState` is a BLE library enum, and it is not exposed outside of the *lib_provisioner* module.
By using a local state and mapping the "internal state" to one that is public, the issue is solved. That was my reasoning for keeping the `ConnectionStatus`.

File: lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/internal/response/ResponsePacket.kt (hunk 6 above)
Comment (philips77) [reply lib_provisioner/src/main/java/no/nordicsemi/wifi/provisioner/library/internal/response/ResponsePacket.kt:46]: Please, check the latest changes.

`suspend...()` methods can throw bunch of different exceptions, not only `InvalidDataException`. Also, when the device gets disconnected, Bluetooth is disconnected, etc. Before, all those were re-thrown from the provisioning library, but catching them would require the BLE library. Now, all of them are mapped to `ResponseErrorException(ResponseError.INTERNAL_ERROR)`. I don't know should there be a new exception type defined...

**Pull request closed without merging**